File: solutions/sonnet-1/sonnet/typescript/rle.ts
const input = await Bun.stdin.text();
const line = input.split("\n")[0].replace(/\r$/, "");

let output = "";
let i = 0;
while (i < line.length) {
  const ch = line[i];
  let count = 0;
  while (i < line.length && line[i] === ch) {
    count++;
    i++;
  }
  output += ch + count;
}

console.log(output);

File: solutions/sonnet-1/sonnet/rust/rle.rs
use std::io::{self, Read, Write};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let line = input.lines().next().unwrap_or("");

    let mut output = String::new();
    let mut chars = line.chars().peekable();
    while let Some(c) = chars.next() {
        let mut count = 1;
        while chars.peek() == Some(&c) {
            chars.next();
            count += 1;
        }
        output.push(c);
        output.push_str(&count.to_string());
    }

    let stdout = io::stdout();
    let mut handle = stdout.lock();
    writeln!(handle, "{}", output).unwrap();
}
